source/common/router/router.cc

```diff
@@ -219,8 +220,23 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
       return Http::FilterHeadersStatus::StopIteration;
     }
     config_.stats_.rq_direct_response_.inc();
-    sendLocalReply(route_->directResponseEntry()->responseCode(), "", false);
-    // TODO(brian-pane) support sending a response body and response_headers_to_add.
+    std::string body;
+    Optional<std::string> inline_body = route_->directResponseEntry()->responseBody();
+    if (inline_body.valid()) {
+      body = inline_body.value();
+    }
+    Optional<std::string> path = route_->directResponseEntry()->responseBodyFilename();
+    if (path.valid()) {
+      const ssize_t kMaxFileSize = 4096;
+      ssize_t file_size = Filesystem::fileSize(path.value());
+      if (file_size < 0 || file_size > kMaxFileSize) {
```

Comment:
While file size can change (so totally check here) it's probably worth best-effort validation of file size at config load time.  Also is the limitation documented anywhere?

Reply:
I'm planning to add documentation for the feature with a follow-up PR that adds the one remaining part of the implementation: support for sending the `response_headers_to_add` along with direct responses like these.